### PR TITLE
Added dry run support.

### PR DESCRIPTION
**Pull request by patriksvensson**
By providing the parameter `dryrun`, `noop` or `whatif`, a [dry run](http://en.wikipedia.org/wiki/Dry_run_%28testing%29) will be performed. This is useful to see what tasks will be executed given the current context (environment variables, arguments etc). 

No tasks will actually be executed.

```
PS> .\Cake.exe build.cake -dryrun

Performing dry run...
Target is: Build

1. Clean
2. Build
3. Copy-Files
4. Package-Files

This was a dry run.
No tasks were actually executed.
```

Related to issue #155.

### CODE REVIEW

**Comment by devlead** [comment]
@patriksvensson will test some more, but seems to works well, should add to Cake.exe usage too before merge.

**Comment by patriksvensson** [comment]
OK, will add to usage. Hold on :+1:

File: src/Cake.Core.Tests/Unit/CakeEngineTests.cs [hunk 1]
```diff
@@ -96,7 +96,8 @@ public void Should_Throw_If_Target_Is_Null()
                     var engine = fixture.CreateEngine();
 
                     // When
-                    var result = Record.Exception(() => engine.RunTarget(fixture.Context, null));
+                    var result = Record.Exception(() =>
```
Comment (devlead) [inline src/Cake.Core.Tests/Unit/CakeEngineTests.cs]: Remove trailing spaces or newline.

File: src/Cake.Core.Tests/Unit/CakeEngineTests.cs [hunk 2]
```diff
@@ -113,7 +114,8 @@ public void Should_Throw_If_Target_Is_Null()
                     var engine = fixture.CreateEngine();
 
                     // When
-                    var result = Record.Exception(() => engine.RunTarget(fixture.Context, new DefaultExecutionStrategy(), null));
+                    var result = Record.Exception(() =>
```
Comment (devlead) [inline src/Cake.Core.Tests/Unit/CakeEngineTests.cs]: Remove trailing spaces or newline.

File: src/Cake.Core/CakeEngine.cs [hunk 3]
```diff
@@ -142,20 +131,11 @@ public CakeReport RunTarget(ICakeContext context, IExecutionStrategy strategy, s
...
-                        _log.Information("========================================");
-                        _log.Information(task.Name);
-                        _log.Information("========================================");
-                        _log.Verbose("Executing task: {0}", task.Name);
-
-                        // Execute the task.
-                        ExecuteTask(context, strategy, stopWatch, task, report);
-
-                        _log.Verbose("Finished executing task: {0}", task.Name);
+                        ExecuteTask(context, strategy, stopWatch, task, report);
```
Comment (devlead) [inline src/Cake.Core/CakeEngine.cs]: Trailing spaces.

File: src/Cake.Core/DefaultExecutionStrategy.cs [hunk 4]
```diff
@@ -40,11 +64,31 @@ public void Execute(CakeTask task, ICakeContext context)
...
 
         /// <summary>
+        /// Skips the specified task.
+        /// </summary>
+        /// <param name="task">The task to skip.</param>
+        public void Skip(CakeTask task)
+        {
+            if (task != null)
+            {
+                _log.Verbose("Skipping task: {0}", task.Name);
```
Comment (devlead) [inline src/Cake.Core/DefaultExecutionStrategy.cs]: Trailing spaces.

File: src/Cake.Core/DefaultExecutionStrategy.cs [hunk 5]
```diff
@@ -40,11 +64,31 @@ public void Execute(CakeTask task, ICakeContext context)
...
         /// <summary>
+        /// Skips the specified task.
+        /// </summary>
+        /// <param name="task">The task to skip.</param>
+        public void Skip(CakeTask task)
+        {
+            if (task != null)
+            {
+                _log.Verbose("Skipping task: {0}", task.Name);   
+            }
```
Comment (devlead) [inline src/Cake.Core/DefaultExecutionStrategy.cs]: Trailing spaces.

File: src/Cake.Tests/Unit/Scripting/DescriptionScriptHostTests.cs [hunk 6]
```diff
@@ -22,7 +67,10 @@ public void Should_Not_Call_To_Engine()
                 host.RunTarget("Target");
 
                 // Then
-                engine.Received(0).RunTarget(context, "Target");
+                engine.Received(0).RunTarget(
+                    context,
```
Comment (devlead) [inline src/Cake.Tests/Unit/Scripting/DescriptionScriptHostTests.cs]: Trailing spaces.

File: src/Cake.Tests/Unit/Scripting/DescriptionScriptHostTests.cs [hunk 7]
```diff
@@ -22,7 +67,10 @@ public void Should_Not_Call_To_Engine()
                 host.RunTarget("Target");
 
                 // Then
-                engine.Received(0).RunTarget(context, "Target");
+                engine.Received(0).RunTarget(
+                    context, 
+                    Arg.Any<DefaultExecutionStrategy>(),
```
Comment (devlead) [inline src/Cake.Tests/Unit/Scripting/DescriptionScriptHostTests.cs]: Trailing spaces.

File: src/Cake/Commands/CommandFactory.cs [hunk 8]
```diff
@@ -4,16 +4,20 @@ internal sealed class CommandFactory : ICommandFactory
     {
         private readonly BuildCommand.Factory _buildCommandFactory;
         private readonly DescriptionCommand.Factory _descriptionCommandFactory;
+        private readonly DryRunCommand.Factory _dryRunCommandFactory;
         private readonly HelpCommand.Factory _helpCommandFactory;
         private readonly VersionCommand.Factory _versionCommandFactory;
 
-        public CommandFactory(BuildCommand.Factory buildCommandFactory,
```
Comment (devlead) [inline src/Cake/Commands/CommandFactory.cs]: Trailing spaces.

File: src/Cake/Commands/ICommandFactory.cs [hunk 9]
```diff
@@ -27,6 +33,6 @@ public interface ICommandFactory
         /// Creates the version command.
         /// </summary>
         /// <returns>The version command.</returns>
-        ICommand CreateVersionCommand();
+        ICommand CreateVersionCommand();
```
Comment (devlead) [inline src/Cake/Commands/ICommandFactory.cs]: Trailing spaces.

File: src/Cake/Scripting/BuildScriptHost.cs [hunk 10]
```diff
@@ -16,10 +18,15 @@ public sealed class BuildScriptHost : ScriptHost
         /// <param name="engine">The engine.</param>
         /// <param name="context">The context.</param>
         /// <param name="reportPrinter">The report printer.</param>
-        public BuildScriptHost(ICakeEngine engine, ICakeContext context, ICakeReportPrinter reportPrinter)
-            : base(engine, context)
+        /// <param name="log">The log.</param>
+        public BuildScriptHost(
+            ICakeEngine engine,
```
Comment (devlead) [inline src/Cake/Scripting/BuildScriptHost.cs]: Trailing spaces.

File: src/Cake/Scripting/BuildScriptHost.cs [hunk 11]
```diff
@@ -16,10 +18,15 @@ public sealed class BuildScriptHost : ScriptHost
         /// <param name="engine">The engine.</param>
         /// <param name="context">The context.</param>
         /// <param name="reportPrinter">The report printer.</param>
-        public BuildScriptHost(ICakeEngine engine, ICakeContext context, ICakeReportPrinter reportPrinter)
-            : base(engine, context)
+        /// <param name="log">The log.</param>
+        public BuildScriptHost(
+            ICakeEngine engine, 
+            ICakeContext context,
```
Comment (devlead) [inline src/Cake/Scripting/BuildScriptHost.cs]: Trailing spaces.

File: src/Cake/Scripting/DryRunExecutionStrategy.cs [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+        public void PerformTeardown(Action action)
+        {
+        }
+
+        public void Execute(CakeTask task, ICakeContext context)
+        {
+            if (task != null)
+            {
+                _log.Information("{0}. {1}", _counter, task.Name);
+                _counter++;
```
Comment (devlead) [inline src/Cake/Scripting/DryRunExecutionStrategy.cs]: Trailing spaces.

File: src/Cake/Scripting/DryRunScriptHost.cs [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+    {
+        private readonly ICakeLog _log;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="DryRunScriptHost"/> class.
+        /// </summary>
+        /// <param name="engine">The engine.</param>
+        /// <param name="context">The context.</param>
+        /// <param name="log">The log.</param>
+        public DryRunScriptHost(ICakeEngine engine, ICakeContext context, ICakeLog log)
```
Comment (devlead) [inline src/Cake/Scripting/DryRunScriptHost.cs]: Trailing spaces.

**Pull request merged**